### PR TITLE
Timebar refactor

### PR DESCRIPTION
**Pull request by j8seangel**
# FEATURES CHANGELOG

- Use distinct icons/shapes for event types both on timebar and map. Harmonized colors (only use white for highlight/hover, fishing events using track color, better sorting out between tracks)
Screenshot 2022-01-25 at 13 29 15

- Better timebar hover inspection: shows events, speed, elevation for multiple tracks at the same time. Inspect values for activity graphs
Screenshot 2022-01-25 at 12 13 09
Screenshot 2022-03-07 at 11 02 41

- Better timebar hover inspection: allows highlighting multiple events on the map at the same time
https://user-images.githubusercontent.com/1583415/150977458-4d3b737a-995b-4e10-aa29-8ba563847ccb.mov

- Simplified chart selector 

- Chart is now available for environment layers
Screenshot 2022-03-07 at 11 01 25

- Added more last period shortcuts 

- Added elevation chart
Screenshot 2022-01-25 at 13 32 18

- Added ability to show speed/elevation charts for two tracks at the same time
Screenshot 2022-01-25 at 13 33 17

- Improved speed/elevation charts rendering by properly calculating max height
Screenshot 2022-01-25 at 13 33 58

- Month interval now available for activity layers

- Better "magnetism/stick to unit" behavior: when changing time selection, it will round selected values to nearest hour/day/month depending on what's available on the currently visible layers (ie what's shown in the timebar will reflect exactly what's shown on the map)

# TODO as of February

- [x] Fix Missing vessel name when hovering on a cluster (there is no track at this point)
- [x] Loitering events rendering (hollow losange) in timebar, map, legend
- [x] Fishing events should always be behind other event types on map (separate into 2 distincts layers each with different z order group)
- [x] Fix issues when switching between activity graph and vessel graphs
- [ ] Review/fix max graph height calculation

# TODO
- [x] Define common data type
- [x] Performance
  - [x] generic filtering hook
  - [x] generic clustering hook
- [ ] graphs refacto (use common data type, convert to TS, cleanup, performanche checks)
  - [x] tracks
  - [x] track events
  - [x] track speed/elevation
  - [x] fishing activity
  - [ ] timeline units
  - [x] highlighter (use generic data struct)
  - [x] centralize derived data generation (Recoil)
 - [ ] visualisation 
   - [x] track events symbols https://github.com/GlobalFishingWatch/frontend/pull/951
   - [x] speed up+down/elevation down
   - [x] put back "compare graphs" mode (2 graphs in // up or down)
   - [x] properly calculate speed/elevation graph max height
   - [ ] graph per latitude/longitude?
   - [ ] https://docs.google.com/presentation/d/10gwrmnTHDi1b123Ly-WiCAG-L4vcAfXZTh2QNgQF1MI/edit#slide=id.geac5be4c86_0_9
- [ ] Timebar UI
  - [x] last xxx days etc buttons 
  - [ ] custom date pickers?
  - [ ] "zoom level" indicator
  - [x] simplified timebar mode selector
    - [x] icons
    - [x] radio disabled/unselected styles
- [ ] Misc. fixes
  - [x] hide speed/ele graph when loading 2 tracks   
  - [x] loading states for graph modes and activity mode
  - [x] precision/chunks (lock drag to chunks?)
  - [x] absolute tracks max
  - [ ] fix or remove blurry thing

### CODE REVIEW

File: libs/timebar/src/chartsNew/track-graph.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    if (!trackGraphData.chunks.length) return 0
+    const itemValues = trackGraphData.chunks.reduce(
+      (acc: number[], currentChunk: TimebarChartDataChunk) => {
+        const chunkValues = currentChunk.values!.map((v) => v.value as number)
+        return acc.concat(chunkValues)
+      },
+      []
+    )
+
+    // https://online.stat.psu.edu/stat200/lesson/3/3.2
```
Comment (j8seangel) [inline libs/timebar/src/chartsNew/track-graph.tsx:22]: 🙇 🥇

File: libs/timebar/src/chartsNew/track-graph.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+const getPaths = (
+  trackGraphData: TimebarChartDataItem,
+  numTracks: number,
+  trackIndex: number,
+  graphHeight: number,
+  overallScale: TimelineScale,
+  maxValue: number,
+  orientation: string
```
Comment (j8seangel) [inline libs/timebar/src/chartsNew/track-graph.tsx:41]: I'd type the possible orientation options

**Review by j8seangel — COMMENTED** [review]
Thanks for all the changes, it is working nice! Lot of comments trying to polish below 👇

File: apps/fishing-map/features/timebar/Timebar.tsx [hunk 3]
```diff
@@ -190,14 +189,41 @@ const TimebarWrapper = () => {
...
+  }, [timebarGraph, tracksGraphsData])
 
-  const highlighterActivity = useMemo(() => {
-    return timebarVisualisation === TimebarVisualisations.Vessel && showGraph && tracksGraphs
-      ? tracksGraphs
-      : null
-  }, [timebarVisualisation, showGraph, tracksGraphs])
+  const highlighterData = useMemo(() => {
+    if (timebarVisualisation === TimebarVisualisations.Heatmap) {
+      // TODO move setStackedActivity from TimebarActivityGraph to a hook that will be called at this level
```
[the remark this answers is not in the document]
Comment (nerik) [reply apps/fishing-map/features/timebar/Timebar.tsx:211]: 👍

File: apps/fishing-map/features/timebar/Timebar.tsx [hunk 4]
```diff
@@ -190,14 +189,41 @@ const TimebarWrapper = () => {
...
+    )
+  }, [timebarGraph, tracksGraphsData])
+
+  const trackGraphOrientation = useMemo(() => {
+    if (tracksGraphsData.length === 0 || tracksGraphsData.length > 2) return 'middle'
+    return {
+      none: 'middle',
+      speed: 'middle',
+      elevation: 'down',
+    }[timebarGraph]
```
[the remark this answers is not in the document]
Comment (nerik) [reply apps/fishing-map/features/timebar/Timebar.tsx]: https://github.com/GlobalFishingWatch/frontend/pull/1213/commits/ff8177bc5b1f43fad2cf9e853b538e0ddd66fe7a

File: apps/fishing-map/features/timebar/Timebar.tsx [hunk 5]
```diff
@@ -229,9 +255,13 @@ const TimebarWrapper = () => {
...
                 <Fragment>
-                  <TimebarTracks key="tracks" tracks={tracks} />
-                  {showGraph && tracksGraphs && (
-                    <TimebarActivity key="trackActivity" graphTracks={tracksGraphs} />
+                  <TimebarTracks key="tracks" data={tracks} orientation={trackGraphOrientation} />
+                  {showGraph && tracksGraphsData && (
+                    <TimebarTrackGraph
+                      key="trackGraph"
+                      data={tracksGraphsData}
+                      orientation={trackGraphOrientation}
```
[the remark this answers is not in the document]
Comment (nerik) [reply apps/fishing-map/features/timebar/Timebar.tsx]: https://github.com/GlobalFishingWatch/frontend/pull/1213/commits/ff8177bc5b1f43fad2cf9e853b538e0ddd66fe7a

File: libs/timebar/src/chartsNew/track-graph.tsx (hunk 2 above)
Comment (nerik) [reply libs/timebar/src/chartsNew/track-graph.tsx]: https://github.com/GlobalFishingWatch/frontend/pull/1213/commits/ff8177bc5b1f43fad2cf9e853b538e0ddd66fe7a

File: libs/timebar/src/immediateContext.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+import { createContext } from 'react'
+
+export const ImmediateContext = createContext<boolean>(false)
```
[the remark this answers is not in the document]
Comment (nerik) [reply libs/timebar/src/immediateContext.ts]: Thanks 👍

File: libs/timebar/src/chartsNew/common/utils.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+export const TOP_MARGIN = 5
+export const BOTTOM_MARGIN = 20
+export const MIN_HEIGHT = 2
+export const TOTAL_MARGIN = TOP_MARGIN + BOTTOM_MARGIN
+
+export const getTrackY = (numTracks: number, trackIndex: number, graphHeight: number) => {
+  const TOTAL_MARGIN = 15
```
[the remark this answers is not in the document]
Comment (nerik) [reply libs/timebar/src/chartsNew/common/utils.ts]: https://github.com/GlobalFishingWatch/frontend/pull/1213/commits/c4d950c3bb3593770604d7612ac17aacb478f4ad

File: libs/timebar/src/immediateContext.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { createContext } from 'react'
+
+export type ImmediateContextProps = {
+  immediate: boolean
+  toggleImmediate: () => void
+}
+export const ImmediateContext = createContext<ImmediateContextProps>({
```
[the remark this answers is not in the document]
Comment (nerik) [reply libs/timebar/src/immediateContext.ts:7]: thanks 👍

File: libs/timebar/src/chartsNew/tracks.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
+import React, { Fragment, useContext, useMemo } from 'react'
+import { ResourceStatus } from '@globalfishingwatch/api-types'
```
[the remark this answers is not in the document]
Comment (nerik) [reply libs/timebar/src/chartsNew/tracks.tsx:2]: @j8seangel This doesn't work with enums 🤮

File: libs/timebar/src/chartsNew/tracks.tsx [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+import TimelineContext, { TimelineScale } from '../timelineContext'
+import ImmediateContext from '../immediateContext'
+import { DEFAULT_CSS_TRANSITION } from '../constants'
+import { getTrackY } from './common/utils'
+import styles from './tracks.module.css'
+import { useFilteredChartData } from './common/hooks'
+import { TimebarChartData, TimebarChartDataChunk, TimebarChartDataItem } from '.'
+
+type TrackChunkWithCoords = TimebarChartDataChunk & { x?: number; width?: number }
+type TrackItemWithCoords = Omit<TimebarChartDataItem, 'chunks'> & {
```
[the remark this answers is not in the document]
Comment (nerik) [reply libs/timebar/src/chartsNew/tracks.tsx]: Ended up adding x and width to the main type

**Pull request merged**